### PR TITLE
Add command to elevate a user (or the bot) as room administrator

### PR DESCRIPTION
**Pull request by maranda**
This is a rebase of #126. I added tests yet as of now only the first self test works the second never does, fwiw I passed hours checking the correctness of the code (tried dozens and dozens of variants) and that there's not an error on my part.
The problem always happen when getFirstReaction() gets called on the second test, the message gets sent correctly in the management room, yet Mjolnir doesn't seem to pick it up.. So the problem looks to be into the mx-tester's boilerplate at this rate.

### CODE REVIEW

File: src/commands/MakeRoomAdminCommand.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+/*
+Copyright 2021 Marco Cirillo
```
[the remark this answers is not in the document]
Comment (Yoric) [reply src/commands/MakeRoomAdminCommand.ts:2]: Also, the copyright should be transferred to the Matrix Foundation. Otherwise, it's by default illegal for us to accept further contributions on this file.

File: src/commands/MakeRoomAdminCommand.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+limitations under the License.
+*/
+
+import { Mjolnir } from "../Mjolnir";
+import { RichReply } from "matrix-bot-sdk";
+
+// !mjolnir make admin <room> [<user ID>]
+export async function execMakeRoomAdminCommand(roomId: string, event: any, mjolnir: Mjolnir, parts: string[]) {
+    const isAdmin = await mjolnir.isSynapseAdmin();
+    if (!isAdmin) {
```
Comment (Yoric) [inline src/commands/MakeRoomAdminCommand.ts:23]: Could you also add a setting to enable/disable this feature?
You'll need to patch `config.ts`, `default.yaml` and `harness.yaml`.

Let's create a setting

```ts
admin?: {
   makeRoomAdminCommand: boolean;
}
```

I'm sure that we'll want to use this `admin` object for other reasons.

File: test/integration/commands/makedminCommandTest.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            await getFirstReaction(mjolnir, this.mjolnir.managementRoomId, '✅', async () => {
+                LogService.debug("makeadminTest", `Sending: !mjolnir make admin ${targetRoom}`);
+                return await moderator.sendMessage(this.mjolnir.managementRoomId, { msgtype: 'm.text', body: `!mjolnir make admin ${targetRoom}` });
+            });
+        } finally {
+            await moderator.stop();
+        }
+        LogService.debug("makeadminTest", `Making self admin`);
+
+        assert.ok(await mjolnir.userHasPowerLevelForAction(mjolnirUserId, targetRoom, PowerLevelAction.Ban), "Bot user is now room admin.");
```
Comment (Yoric) [inline test/integration/commands/makedminCommandTest.ts:40]: Just to be on the safe side, could you add a check that the bot is not admin yet before the call to `getFirstReaction`?

File: test/integration/commands/makedminCommandTest.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    it('Mjölnir make the bot self room administrator', async function () {
+        this.timeout(60000);
+        const mjolnir = config.RUNTIME.client!;
+        const mjolnirUserId = await mjolnir.getUserId();
+        const moderator = await newTestUser({ name: { contains: "moderator" } });
+        this.moderator = moderator;
+
+        await moderator.joinRoom(config.managementRoom);
+        LogService.debug("makeadminTest", `Joining managementRoom: ${config.managementRoom}`);
+        let targetRoom = await moderator.createRoom({ invite: [mjolnirUserId] });
```
Comment (Yoric) [inline test/integration/commands/makedminCommandTest.ts:25]: Could you also invite another user? We'll test that the user is *not* made admin.

File: test/integration/commands/makedminCommandTest.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    it('Mjölnir make the tester room administrator', async function () {
+        this.timeout(60000);
+        const mjolnir = config.RUNTIME.client!;
+        const moderator = await newTestUser({ name: { contains: "moderator" } });
+        const userA = await newTestUser({ name: { contains: "a" } });
+        const userB = await newTestUser({ name: { contains: "b" } });
+        const userBId = await userB.getUserId();
+        this.moderator = moderator;
+        this.userA = userA;
+        this.userB = userB;
```
Comment (Yoric) [inline test/integration/commands/makedminCommandTest.ts:52]: Likewise, could you add a userC (as control sample) who won't be made admin?

**Review by Yoric — COMMENTED** [review]
Thanks a lot for the PR!

I have a few nitpicks, mostly on the test.

File: src/commands/MakeRoomAdminCommand.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (maranda) [reply src/commands/MakeRoomAdminCommand.ts]: Added sign out to commits.

File: src/commands/MakeRoomAdminCommand.ts (hunk 2 above)
Comment (maranda) [reply src/commands/MakeRoomAdminCommand.ts:23]: I can do that, but considering it's a command and not some protection that is pluggable am I supposed to hide it from the command list and do conditional loading of the import in the command handler if it's not activated in the config?

File: test/integration/commands/makedminCommandTest.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        await moderator.joinRoom(this.mjolnir.managementRoomId);
+        LogService.debug("makeadminTest", `Joining managementRoom: ${this.mjolnir.managementRoomId}`);
+        let targetRoom = await userA.createRoom({ invite: [userBId] });
+        LogService.debug("makeadminTest", `User A creating targetRoom: ${targetRoom}; and inviting ${userBId}`);
+        try {
+            await userB.start();
+            userB.joinRoom(targetRoom);
+        } finally {
+            LogService.debug("makeadminTest", `${userBId} joining targetRoom: ${targetRoom}`);
+            await userB.stop();
```
[the remark this answers is not in the document]
Comment (maranda) [reply test/integration/commands/makedminCommandTest.ts:70]: I need userB only to join, the actual looking glass for the powerlevel is done by userA.

File: test/integration/commands/makedminCommandTest.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            await getFirstReaction(mjolnir, this.mjolnir.managementRoomId, '✅', async () => {
+                LogService.debug("makeadminTest", `Sending: !mjolnir make admin ${targetRoom} ${userBId}`);
+                return await moderator.sendMessage(this.mjolnir.managementRoomId, { msgtype: 'm.text', body: `!mjolnir make admin ${targetRoom} ${userBId}` });
+            });
+        } finally {
+            await moderator.stop();
+        }
+        LogService.debug("makeadminTest", `Making User B admin`);
+
+        await assert.ok(await userA.userHasPowerLevelForAction(userBId, targetRoom, PowerLevelAction.Ban), "User B is now room admin.");
```
[the remark this answers is not in the document]
Comment (maranda) [reply test/integration/commands/makedminCommandTest.ts]: See above comment.

File: test/integration/commands/makedminCommandTest.ts (hunk 3 above)
Comment (maranda) [reply test/integration/commands/makedminCommandTest.ts]: Done.

File: test/integration/commands/makedminCommandTest.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (maranda) [reply test/integration/commands/makedminCommandTest.ts]: Done.

File: test/integration/commands/makedminCommandTest.ts (hunk 5 above)
Comment (maranda) [reply test/integration/commands/makedminCommandTest.ts:68]: Done.

File: test/integration/commands/makedminCommandTest.ts (hunk 4 above)
Comment (maranda) [reply test/integration/commands/makedminCommandTest.ts:30]: Done.

File: src/commands/MakeRoomAdminCommand.ts (hunk 2 above)
Comment (Yoric) [reply src/commands/MakeRoomAdminCommand.ts:23]: That would be the ideal case, but I feel that this is beyond the scope of your PR. One of these days, we'll refactor commands to make them easier to add/remove.

For the time being, what about rejecting from within `execMakeRoomAdminCommand`, just as you do when `!isAdmin`?

Something like

```ts
if (!config.?admin?.enableMakeRoomAdminCommand) {
   // ...reject
}
```

File: config/default.yaml [hunk 8]
```diff
@@ -95,6 +95,14 @@ protectedRooms:
 # Manually add these rooms to the protected rooms list if you want them protected.
 protectAllJoinedRooms: false
 
+# Server administration commands
+admin:
+  # The make admin command allows you to use the Synapse admin API to force
+  # a room dministrator of the target room which is local to Mjolnir's HomeServer
```
[the remark this answers is not in the document]
Comment (Yoric) [reply config/default.yaml:101]: Nit: `make admin` (between backticks), to clarify that it's the name of the command.
Also, I would rephrase along the lines of

```
The `make admin` upgrades the powerlevel of a specified user (or the bot itself) of a room to make them admin of the room (powerlevel 100).

This only works if the room has at least one admin on the local homeserver (the homeserver specified in `homeserverUrl` in this file).
```

**Review by Yoric — COMMENTED** [review]
We're getting quite close!

File: src/commands/MakeRoomAdminCommand.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    const isAdmin = await mjolnir.isSynapseAdmin();
+    if (!config.admin?.enableMakeRoomAdminCommand || !isAdmin) {
+        const message = "The command is either not enabled in the config, or I can't perform the requested operation";
+        const reply = RichReply.createFor(roomId, event, message, message);
+        reply['msgtype'] = "m.notice";
+        mjolnir.client.sendMessage(roomId, reply);
+        return;        
+    }
+
+    await mjolnir.makeUserRoomAdmin(await mjolnir.client.resolveRoom(parts[3]), parts[4]);
```
[the remark this answers is not in the document]
Comment (maranda) [reply src/commands/MakeRoomAdminCommand.ts]: Done with https://github.com/matrix-org/mjolnir/pull/219/commits/189c874880f0418b43b56f15590bf9bfe7413b29

[the remark this answers is not in the document]
Comment (maranda) [reply src/commands/MakeRoomAdminCommand.ts]: And it's a Bad Request (400): {"errcode":"M_UNKNOWN","error":"No local admin user in room"}

**Comment by Yoric** [comment]
@maranda When you're ready for another review, don't forget to ping me :)

**Comment by maranda** [comment]
@Yoric ready :)

**Review by Yoric — APPROVED** [review]
Looks good to me, thanks a lot!

Ready to merge when you are.

**Comment by maranda** [comment]
@Yoric 👌
